core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -139,43 +143,51 @@ protected void initChannel(Channel ch) throws Exception {
+               sslCtxHttp1Only : sslCtxHttp1Or2;
     }
 
-    private void configureProxy(Channel ch, ProxyConfig proxyConfig, SslContext sslCtx) {
+    private void configureProxy(Channel ch, ProxyConfig proxyConfig, SessionProtocol desiredProtocol) {
+        if (proxyConfig.proxyType() == ProxyType.DIRECT) {
+            return;
+        }
         final ProxyHandler proxyHandler;
+        final InetSocketAddress proxyAddress = proxyConfig.proxyAddress();
+        assert proxyAddress != null;
         switch (proxyConfig.proxyType()) {
-            case DIRECT:
-                return;
             case SOCKS4:
                 final Socks4ProxyConfig socks4ProxyConfig = (Socks4ProxyConfig) proxyConfig;
-                proxyHandler = new Socks4ProxyHandler(socks4ProxyConfig.proxyAddress(),
-                                                      socks4ProxyConfig.username());
+                proxyHandler = new Socks4ProxyHandler(proxyAddress, socks4ProxyConfig.username());
                 break;
             case SOCKS5:
                 final Socks5ProxyConfig socks5ProxyConfig = (Socks5ProxyConfig) proxyConfig;
-                proxyHandler = new Socks5ProxyHandler(
-                        socks5ProxyConfig.proxyAddress(), socks5ProxyConfig.username(),
-                        socks5ProxyConfig.password());
+                proxyHandler = new Socks5ProxyHandler(proxyAddress, socks5ProxyConfig.username(),
+                                                      socks5ProxyConfig.password());
                 break;
             case CONNECT:
                 final ConnectProxyConfig connectProxyConfig = (ConnectProxyConfig) proxyConfig;
                 final String username = connectProxyConfig.username();
                 final String password = connectProxyConfig.password();
                 if (username == null || password == null) {
-                    proxyHandler = new HttpProxyHandler(connectProxyConfig.proxyAddress());
+                    proxyHandler = new HttpProxyHandler(proxyAddress);
                 } else {
-                    proxyHandler = new HttpProxyHandler(connectProxyConfig.proxyAddress(), username, password);
-                }
-                if (connectProxyConfig.useTls()) {
-                    ch.pipeline().addLast(sslCtx.newHandler(ch.alloc()));
+                    proxyHandler = new HttpProxyHandler(proxyAddress, username, password);
                 }
                 break;
             default:
                 throw new Error(); // Should never reach here.
         }
         proxyHandler.setConnectTimeoutMillis(connectTimeoutMillis);
-        ch.pipeline().addLast(proxyHandler);
+        ch.pipeline().addFirst(proxyHandler);
+
+        if (proxyConfig instanceof ConnectProxyConfig && ((ConnectProxyConfig) proxyConfig).useTls()) {
+            final SslContext sslCtx = determineSslContext(desiredProtocol);
+            ch.pipeline().addFirst(sslCtx.newHandler(ch.alloc()));
```

Comment:
Question: What happens if `sslCtx` is applied twice? here and in `HttpClientPipelineConfigurator`?

Reply:
Nothing interesting 😄  just ssl is applied twice.
Let me add an integration test to make sure this behavior isn't broken in the future though

Reply:
This is embarrassing... but found a bug thanks to your comment 👍 
Let me handle this issue in a separate PR 🙇